README
Skribilo
========

Skribilo is a document authoring system allowing users to produce
documents in a variety of formats (HTML, PostScript, PDF, XML, etc.)
from a single source document, while still retaining fine-grain control
over the output.

Beside rudimentary features provided by the `base' package, Skribilo
provides packages for the production of colored program source code,
bibliographies, pie charts, mathematical equations, and more.

Skribilo is based on the Scheme programming language and is highly
extensible: it is easy to define new documentation primitives (called
"markups") and their rendering in each output format.

Skribilo's core is written in mostly-portable Scheme (i.e., uses
standard APIs such as SRFIs rather than Guile-specific APIs whenever
that is possible), making it intelligible and hackable for most Scheme
programmers.

For more information, see:

  http://www.nongnu.org/skribilo/


Requirements
............

You need at least:

  * GNU Guile >= 2.0.x
    http://www.gnu.org/software/guile/

  * Guile-Reader >= 0.6
    http://www.nongnu.org/guile-reader/

  * ImageMagick, used to convert images from one format to another
    http://imagemagick.org/

Optionally, you may want to install one of the following document layout
programs to produce PS/PDF files:

  * Lout
    http://lout.sourceforge.net/

  * LaTeX
    http://www.latex-project.org/

  * ConTeXt
    http://www.pragma-ade.nl/

To produce pie charts, you may want to use:

  * Ploticus
    http://ploticus.sourceforge.net/

To convert images in the `fig' format:

  * Xfig, Transfig
    http://xfig.org/

To be able to use the `rss-2' reader, you will need:

  * Guile-Lib (specifically, `sxml' and `htmlprag' modules)
    http://www.nongnu.org/guile-lib/

This is it!


Development
...........

A copy of the source code repository can be obtained with Git
(http://git.or.cz/), using the following incantation:

  git clone git://git.sv.gnu.org/skribilo.git

You can then send patches to the <skribilo-users@nongnu.org> mailing
list!


History
.......

Skribilo is based on the Skribe code base, written by Manuel Serrano and
Erick Gallesio.  See `README.Skribe' for the original Skribe `README'
file, or visit:

  http://www-sop.inria.fr/mimosa/fp/Skribe/

I am thankful to them for implementing Skribe!


June 2007,
Ludovic Courtès <ludo@chbouib.org>.



Copyright 2007, 2008, 2012  Ludovic Courtès <ludo@gnu.org>

  Copying and distribution of this file, with or without modification,
  are permitted in any medium without royalty provided the copyright
  notice and this notice are preserved.

;;; Local Variables:
;;; mode: text
;;; coding: utf-8
;;; End:


NEWS
Copyright (C) 2005, 2006, 2007, 2008, 2009, 2012, 2013, 2016, 2018, 2020 Ludovic Courtès <ludo@gnu.org>

  Copying and distribution of this file, with or without modification,
  are permitted in any medium without royalty provided the copyright
  notice and this notice are preserved.



* New in Skribilo 0.9.5

** Skribilo now runs on Guile 3.0 (in addition to 2.x)
** Guile 1.8 is no longer supported

** ‘document’ markup supports :info-dir-category and :info-dir-entry

These are honored by the Info engine, which uses them as the category
and entry that show up in the Info “dir” file.

** Info engine: number sections and chapters
** Lout engine: recognize and replace more Unicode characters in the input

* New in Skribilo 0.9.4

** Skribilo now runs on Guile 2.2 (in addition to 2.0 and 1.8)
** HTML engine: do not emit <strong> and <big>

It used to do so for titles and slide titles, which interfered with CSS
customization in addition to being so 1990.

** New style for the HTML user manual, rely on CSS

Items such as markup documentation are now rendered as CSS.  The whole
web site and user manual style has been overhauled thanks to Amirouche
Boubekki.

** Improved Lout style of the user manual
** acmproc: "References" chapter is now unnumbered
** New Friulian and Brazilian Portuguese translations

* New in Skribilo 0.9.3

** Skribilo now speaks your native language

Translations are currently available in Danish, French, Hungarian, and
Swedish.  See <https://translationproject.org/domain/skribilo.html>.
Note that the locale also affects the language used in bibliography
entries and similar.

** HTML engine: all the footnotes are emitted for single-file documents

Previously, only the footnotes of the last chapter or section would be
output.

** HTML engine: each footnote is enclosed in <div class="footnote">

Previously, each footnote was surrounded by <br> and </br>.  The above
change should make CSS customization easier.

** HTML engine: footnote HTML is now valid

It used to be that footnote HTML code contained an extraneous </div>,
and nested <a> tags, which is invalid.

** LaTeX engine: paragraphs no longer systematically preceded by \noindent
** acmproc: fix typos in the LaTeX engine customs
** acmproc: `references' uses a chapter, the default font, and sorts entries
** lncs: style of bibliography entries more closely matches LNCS style.
** `ref' warnings now always print the location of erroneous `:bib' references
** `numref' inserts an unbreakable space (was a breakable space) after its `:text'

* New in Skribilo 0.9.2

** Skribilo is now distributed under GPLv3 or later

Previous versions were under of GPLv2 or later.

** Improved configure-time diagnostics
** `skribilo' now displays a call stack trace upon error when possible
** `skribilo' now installs the user-specified locale
** New `--custom' compiler option
** Using `(image :url ...)' with `lout' yields a warning, not an error
** New `rss-2' input syntax, for RSS 2.0 feeds
** New Info engine, to produce `.info' files browseable with Emacs
** New `:info-node' option for `chapter', `section', etc.
** HTML engine: footnotes now yield hyperlinks back to the call site
** HTML engine: footnotes are properly printed for same-file chapters

Previously, footnotes would not get printed in the case of chapters that
are not output in a separate file.

** HTML engine: use ". " as the default chapter number-title separator
** HTML engine: the `charset' custom changes the output encoding (on Guile 2.0)
** Lout/LaTeX engine: new `encoding' custom
** Lout engine: `lout-illustration' calls the `inline-definitions-proc' custom
** Improved rendering of pie charts with Ploticus
** On Guile 2.0, modules are compiled and `.go' files are installed
** Improve internationalization of informative and error messages
** New test suite (run "make check")
** Various documentation fixes, improvements, and additions

* New in Skribilo 0.9.1

** Improved `configure' diagnostics and build/installation fixes

Thanks to Barrie Stott!

** Various documentation fixes, improvements and additions

Notably, examples no longer use the legacy Skribe API and the ConTeXt
engine is now documented.

** Emacs mode no longer depends on Bigloo UDE
** Guile warnings about duplicate bindings are now hidden

These harmless warnings looked like:

  WARNING: (skribilo ast): imported module (srfi srfi-34) overrides core binding `raise'

** The `slide' package supports additional customs

Namely, `slide-outline-title' and
`slide-outline-{active,inactive}-color'.  See manual for details.

** The `:toc' option of `slide' determines whether to include it in the outline
** New option `:rulecolor' for `table'
** New `--sui-path' command-line option, new `*sui-path*' parameter
** Fix SUI `:ident' references
** Use more SRFI-34/35 exception handling instead of `skribe-error'
** Lesser reliance on recursive make


* New in Skribilo 0.9.0 (user-visible changes compared to Skribe 1.2d)

** New engine: `lout' (see http://lout.sf.net/).

** New packages

  - `eq' for equation formatting.
  - `pie' for pie charts.
  - `diff' for document diff (experimental).
  - `web-book2' for on-line publishing.

** New markups in the `base' package

  `~', `breakable-space', `numref', `!lout', `lout-illustration'.

** Extended markups in the `base' package

   - `footnote' now takes a `:label' option.
   - `document' now takes a `:keywords' option.	
   - `bibliography' now takes a `:labels' option.

** New `outline' front-end (see the `--reader' option of `skribilo')

** New module system (actually that of GNU Guile), no `skribe-load'

  Just like modules help structure programs, they are useful for
  documents: Skribilo documents are not second-class programs.  Also,
  by default, Skribilo documents have fewer bindings in their name
  space than Skribe documents.

  For compatibility with Skribe, see the `--compat=skribe'
  command-line option.

** Little or no global state


src/guile/skribilo.scm
;;; skribilo.scm  --  The Skribilo document processor.
;;;
;;; Copyright 2005, 2006, 2007, 2008, 2009, 2011, 2012, 2013,
;;;   2015, 2016, 2018, 2020  Ludovic Courtès <ludo@gnu.org>
;;; Copyright 2003, 2004  Erick Gallesio - I3S-CNRS/ESSI <eg@unice.fr>
;;;
;;;
;;; This file is part of Skribilo.
;;;
;;; Skribilo is free software: you can redistribute it and/or modify
;;; it under the terms of the GNU General Public License as published by
;;; the Free Software Foundation, either version 3 of the License, or
;;; (at your option) any later version.
;;;
;;; Skribilo is distributed in the hope that it will be useful,
;;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;;; GNU General Public License for more details.
;;;
;;; You should have received a copy of the GNU General Public License
;;; along with Skribilo.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:
;;;
;;; Usage: skribilo [ARGS]
;;;
;;; Process a skribilo document using options from the command-line.
;;;
;;; Code:



(define-module (skribilo)
  #:autoload    (skribilo module) (make-user-module *skribilo-user-module*)
  #:use-module  (skribilo engine)
  #:use-module  (skribilo reader)

  #:use-module  (skribilo utils syntax)
  #:use-module  (skribilo evaluator)
  #:use-module  (skribilo debug)
  #:use-module  (skribilo parameters)
  #:use-module  (skribilo config)

  #:use-module  (srfi srfi-1)
  #:use-module  (srfi srfi-13)
  #:use-module  (srfi srfi-37)
  #:use-module  (srfi srfi-39)

  #:export (skribilo))


;; Install the Skribilo module syntax reader.
(skribilo-module-syntax)

(if (not (keyword? :kw))
    (error "guile-reader sucks"))



;;;
;;; Help.
;;;

(define (skribilo-show-help)
  (display (G_ "Usage: skribilo [OPTIONS] [INPUT]"))
  (display (G_ "
Process Skribilo document from file INPUT (or standard input) using the
specified reader syntax or the 'skribe' syntax, and produce its output using
the specified engine or the 'html' engine.
"))
  (newline)
  (display (G_ "\
  -R, --reader=READER    use READER to parse the input file;
                         e.g., 'skribe' (default) or 'outline'\n"))
  (display (G_ "\
  -t, --target=ENGINE    use ENGINE as the output engine; e.g., 'html'\n"))
  (display (G_ "\
  -c, --custom=C=VAL     use VAL as the value of ENGINE's custom C\n"))
  (display (G_ "\
  -o, --output=FILE      write output to FILE\n"))
  (display (G_ "\
      --compat=COMPAT    use COMPAT as the compatibility layer; e.g., 'skribe'\n"))
  (newline)
  (display (G_ "\
  -I, --doc-path=DIR     prepend DIR to the document include path\n"))
  (display (G_ "\
  -B, --bib-path=DIR     prepend DIR to the bibliography include path\n"))
  (display (G_ "\
  -S, --source-path=DIR  prepend DIR to the source include path\n"))
  (display (G_ "\
  -P, --image-path=DIR   prepend DIR to the image include path\n"))
  (display (G_ "\
  -U, --sui-path=DIR     prepend DIR to the Skribe URL Index (SUI) search path\n"))
  (newline)
  (display (G_ "\
  -b, --base=BASE        strip BASE from all hyperlinks ('html' engine)\n"))
  (display (G_ "\
  -e, --eval=EXPR        prepend EXPR to the list of expressions to be
                         evaluted before the input file is processed\n"))
  (display (G_ "\
  -p, --preload=FILE     preload FILE before processing the input file\n"))
  (newline)
  (display (G_ "\
  -v, --verbose[=LEVEL]  be verbose, unless LEVEL is 0\n"))
  (display (G_ "\
  -w, --warning[=LEVEL]  issue warnings, unless LEVEL is 0\n"))
  (display (G_ "\
  -g, --debug[=ARG]      issue debugging output, unless ARG is 0; if ARG is
                         not a number, it is interpreted as a symbol to be
                         watched\n"))
  (display (G_ "\
      --no-color         disable colored debugging output\n"))
  (newline)
  (display (G_ "\
  -h, --help             display this help text and exit\n"))
  (display (G_ "\
  -V, --version          display version information and exit\n"))
  (newline)
  (format #t (G_ "\
Report bugs to <~a>.~%")
          (skribilo-bug-report-address)))

(define (skribilo-show-version)
  (format #t "skribilo ~a~%" (skribilo-version)))

(define (leave fmt . args)
  (apply format (current-error-port) (string-append fmt "~%") args)
  (exit 1))



;;;
;;; Document processing.
;;;

(define *skribilo-output-port* (make-parameter (current-output-port)))

(define (doskribe module)
  (let ((output-port (current-output-port))
	(user-module (current-module)))
    (dynamic-wind
	(lambda ()
	  ;; FIXME: Using this technique, anything written to `stderr' will
	  ;; also end up in the output file (e.g. Guile warnings).
	  (set-current-output-port (*skribilo-output-port*))
          (set-current-module module)
          (*skribilo-user-module* module))
	(lambda ()
	  ;;(format #t "engine is ~a~%" (*current-engine*))
	  (evaluate-document-from-port (current-input-port)
				       (*current-engine*)))
	(lambda ()
	  (set-current-output-port output-port)
	  (set-current-module user-module)
          (*skribilo-user-module* #f)))))



;;;
;;; Argument parsing.
;;;

(define (make-path-processor key)
  (lambda (opt name arg result)
    (let ((path (assoc key result)))
      (alist-cons key (if (pair? path)
                          (cons arg (cdr path))
                          (list arg))
                  (alist-delete key result eq?)))))

(define (make-level-processor key default)
  (lambda (opt name arg result)
    (alist-cons key (if (string? arg)
                        (or (string->number arg) default)
                        default)
                result)))

(define %options
  ;; Specifications of the command-line options.
  (list (option '(#\h "help") #f #f
                (lambda args
                  (skribilo-show-help)
                  (exit 0)))
        (option '(#\V "version") #f #f
                (lambda args
                  (skribilo-show-version)
                  (exit 0)))

        (option '(#\R "reader") #t #f
                (lambda (opt name arg result)
                  (alist-cons :reader arg result)))
        (option '(#\t "target") #t #f
                (lambda (opt name arg result)
                  (alist-cons :target arg result)))
        (option '(#\c "custom") #t #f
                (lambda (opt name arg result)
                  (let ((=-pos (string-index arg #\=)))
                    (if (not =-pos)
                        (leave (G_ "~a: missing value for custom") arg)
                        (let ((custom (string-take arg =-pos))
                              (value  (string-drop arg (+ =-pos 1))))
                          (catch 'read-error
                            (lambda ()
                              (let ((custom (string->symbol custom))
                                    (value
                                     (with-input-from-string value read))
                                    (result
                                     (alist-delete :customs result eq?))
                                    (customs
                                     (assoc-ref result :customs)))
                                (alist-cons
                                 :customs
                                 (alist-cons custom value customs)
                                 result)))
                            (lambda (key . args)
                              (leave (G_ "~a: invalid custom value")
                                     value))))))))
        (option '(#\o "output") #t #f
                (lambda (opt name arg result)
                  (if (assoc :output result)
                      (leave "~a: only one output at a time" arg)
                      (alist-cons :output arg result))))

        (option '("compat") #t #f
                (lambda (opt name arg result)
                  (alist-cons :compat arg result)))
        (option '(#\I "doc-path") #t #f
                (make-path-processor :doc-path))
        (option '(#\B "bib-path") #t #f
                (make-path-processor :bib-path))
        (option '(#\S "source-path") #t #f
                (make-path-processor :source-path))
        (option '(#\P "image-path") #t #f
                (make-path-processor :image-path))
        (option '(#\U "sui-path") #t #f
                (make-path-processor :sui-path))

        (option '(#\b "base") #t #f
                (lambda (opt name arg result)
                  (if (assoc :ref-base result)
                      (leave "~a: only one reference base at a time" arg)
                      (alist-cons :ref-base arg result))))
        (option '(#\e "eval") #t #f
                (lambda (opt name arg result)
                  (let ((expr-list (assoc :expressions result))
                        (expr      (with-input-from-string arg read)))
                    (alist-cons :expressions
                                (if (pair? expr-list)
                                    (cons expr (cdr expr-list))
                                    (list expr))
                                result))))
        (option '(#\p "preload") #t #f
                (lambda (opt name arg result)
                  (let ((preloads (assoc :preloads result)))
                    (alist-cons :preloads
                                (if (pair? preloads)
                                    (cons arg preloads)
                                    (list arg))
                                result))))

        (option '(#\v "verbose") #f #t
                (make-level-processor :verbose 0))
        (option '(#\w "warning") #f #t
                (make-level-processor :warning 1))
        (option '(#\g "debug") #f #t
                (lambda (opt name arg result)
                  (let ((num (if arg
                                 (string->number arg)
                                 1)))
                    (if (integer? num)
                        (alist-cons :debug num result)
                        (let ((watched (assoc :watched-symbols result)))
                          (alist-cons :watched-symbols
                                      (cons (string->symbol arg)
                                            (cdr watched))
                                      result))))))
        (option '("no-color") #f #f
                (lambda (opt name arg result)
                  (alist-cons :no-color? #t result)))))

(define %default-options
  ;; Default value of various command-line options.
  '((:debug     . 0)
    (:warning   . 1)
    (:verbose   . 0)
    (:reader    . "skribe")
    (:target    . "html")
    (:compat    . "skribilo")
    (:doc-path    ".")
    (:bib-path    ".")
    (:source-path ".")
    (:image-path  ".")
    (:sui-path    ".")
    (:customs)
    (:watched-symbols)))

(define (parse-args args)
  "Parse argument list @var{args} and return an alist with all the relevant
options."
  (args-fold args %options
             (lambda (opt name arg result)
               (leave "~A: unrecognized option" opt))
             (lambda (file result)
               (if (assoc :input result)
                   (leave "~a: only one input file at a time" file)
                   (alist-cons :input file result)))
             %default-options))


;;;
;;; The program.
;;;

(define (skribilo . args)
  ;; Install the user-specified locale.
  (catch 'system-error
    (lambda _
      (setlocale LC_ALL ""))
    (lambda args
      (format (current-error-port)
              (G_ "warning: failed to install locale: ~a~%")
              (strerror (system-error-errno args)))))

  ;; Tell gettext where to look for message catalogs.
  (bindtextdomain %skribilo-text-domain (skribilo-locale-directory))
  (textdomain %skribilo-text-domain)

  (let* ((options           (parse-args args))

	 (reader-name       (string->symbol (assoc-ref options :reader)))
	 (engine            (string->symbol (assoc-ref options :target)))
         (customs           (assoc-ref options :customs))
         (input-file        (assoc-ref options :input))
	 (output-file       (assoc-ref options :output))

         (verbosity-level   (assoc-ref options :verbose))
	 (debugging-level   (assoc-ref options :debug))
	 (warning-level     (assoc-ref options :warning))
         (watched-symbols   (assoc-ref options :watched-symbols))
         (color?            (not (assoc-ref options :no-color?)))

         (ref-base          (assoc-ref options :ref-base))
         (expressions       (assoc-ref options :expressions))

	 (load-path         (assoc-ref options :doc-path))
	 (bib-path          (assoc-ref options :bib-path))
	 (source-path       (assoc-ref options :source-path))
	 (image-path        (assoc-ref options :image-path))
         (sui-path          (assoc-ref options :sui-path))
         (compat            (assoc-ref options :compat))
	 (preloads          (assoc-ref options :preloads))
	 ;;(variants          '()) ;; FIXME: Implement
         )

    (define user-module
      ;; The environment in which the document is evaluated.
      (make-user-module (string->symbol compat)))

    (if (> (*debug*) 4)
	(set! %load-hook
	      (lambda (file)
		(format #t "~~ loading '~a'...~%" file))))


    (parameterize ((*document-reader*   (make-reader reader-name))
		   (*current-engine*    engine)
                   (*ref-base*          ref-base)
		   (*document-path*     load-path)
		   (*bib-path*          bib-path)
		   (*source-path*       source-path)
		   (*image-path*        image-path)
                   (*sui-path*          sui-path)
		   (*debug*             debugging-level)
                   (*debug-use-colors?* color?)
                   (*watched-symbols*   watched-symbols)
		   (*warning*           warning-level)
		   (*verbose*           verbosity-level))

      ;; Load the user rc file (FIXME)
      ;;(load-rc)

      (or (null? customs)
          (let ((engine (lookup-engine engine)))
            (for-each (lambda (custom+value)
                        (let ((custom (car custom+value))
                              (value  (cdr custom+value)))
                          (engine-custom-set! engine custom value)))
                      customs)))

      ;; Evaluate expressions passed as `--eval'.
      (for-each (lambda (expr)
                  (eval expr user-module))
                (or expressions '()))

      ;; Load files passed as `--preload' using the default reader.
      (for-each (lambda (f)
                  (save-module-excursion
                   (lambda ()
                     (set-current-module user-module)
                     (load f))))
		(or preloads '()))

      ;; Load the specified variants. (FIXME)
;;       (for-each (lambda (x)
;; 		  (skribe-load (format #f "~a.skr" x)
;; 			       :engine (*current-engine*)))
;; 		(reverse! variants))

      (if (and output-file (file-exists? output-file))
          (delete-file output-file))

      ;; Choose UTF-8 as the default encoding so that string ports will
      ;; accept all of Unicode.
      (default-to-utf-8
        (parameterize ((*destination-file* output-file)
                       (*source-file*      input-file)
                       (*skribilo-output-port*
                        (if (string? output-file)
                            (open-output-file output-file)
                            (current-output-port))))

          (setvbuf (*skribilo-output-port*)
                   (cond-expand
                     (guile-2.2 'block)           ;Guile 2.2 and 3.0
                     (else      _IOFBF))
                   16384)

          (if input-file
              (with-input-from-file input-file
                (lambda ()
                  (set-correct-file-encoding!)
                  (doskribe user-module)))
              (doskribe user-module))

          ;; Make sure the output port is flushed before we leave.
          (force-output (*skribilo-output-port*)))))))


;;; skribilo ends here.
